### PR TITLE
token wrapper handler fee update

### PR DESCRIPTION
**Pull request by saiakilesh**

### CODE REVIEW

File: pallets/token-wrapper-handler/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+#[test]
+fn should_update_fee() {
+	new_test_ext().execute_with(|| {
+        let src_chain_id = 1;
+		let resource_id = pallet_bridge::utils::derive_resource_id(src_chain_id, b"hash");
+		let prop_id = 1;
+		// create anchor update proposal
+		setup_relayers(src_chain_id);
+        relay_fee_update_proposal(src_chain_id, &resource_id, prop_id, 5);
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/tests.rs:71]: spacing looks off here, maybe `cargo fmt` if you haven't

File: pallets/token-wrapper/src/traits.rs [hunk 2]
```diff
@@ -1,6 +1,9 @@
 use frame_support::dispatch;
 
 pub trait TokenWrapperInterface<AccountId, AssetId, Balance> {
+	fn set_wrapping_fee(
```
Comment (drewstone) [inline pallets/token-wrapper/src/traits.rs:4]: shouldn't need to use line breaks here, it looks like it can be on one line

File: pallets/token-wrapper-handler/src/tests.rs (hunk 1 above)
Comment (saiakilesh) [reply pallets/token-wrapper-handler/src/tests.rs]: fixed

File: pallets/token-wrapper/src/traits.rs (hunk 2 above)
Comment (saiakilesh) [reply pallets/token-wrapper/src/traits.rs]: fixed

File: pallets/token-wrapper-handler/src/lib.rs [hunk 3]
```diff
@@ -42,32 +87,31 @@ pub mod pallet {
...
 
 	#[pallet::error]
 	pub enum Error<T> {
 		/// Access violation.
 		InvalidPermissions,
-		// Anchor handler already exists for specified resource Id.
+		// TokenWrapperHandler already exists for specified resource Id.
 		ResourceIsAlreadyAnchored,
-		// Anchor handler doesn't exist for specified resoure Id.
+		// TokenWrapper handler doesn't exist for specified resoure Id.
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/lib.rs:101]: Spelling

File: pallets/token-wrapper-handler/src/lib.rs [hunk 4]
```diff
@@ -42,32 +87,31 @@ pub mod pallet {
...
 		TokenWrapperHandlerNotFound,
 		/// Storage overflowed.
 		StorageOverflow,
 	}
-
+	/// Execute the wrapping fee proposal by calling the update_wrapping_fee
+	/// method Ensures that only the bridge can call this function
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/lib.rs:107]: These comments should be on top of the function not the #[pallet::call]

File: pallets/token-wrapper-handler/src/lib.rs [hunk 5]
```diff
@@ -78,8 +122,12 @@ pub mod pallet {
 }
 
 impl<T: Config> Pallet<T> {
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/lib.rs:124]: We can remove this entire impl and just use that one line directly in the call.

**Review by drewstone — CHANGES_REQUESTED** [review]

File: pallets/token-wrapper-handler/src/lib.rs [hunk 6]
```diff
@@ -96,17 +95,18 @@ pub mod pallet {
 	pub enum Error<T> {
 		/// Access violation.
 		InvalidPermissions,
-		// TokenWrapperHandler already exists for specified resource Id.
+		// Token Wrapper Handler already exists for specified resource Id.
 		ResourceIsAlreadyAnchored,
-		// TokenWrapper handler doesn't exist for specified resoure Id.
+		// Token Wrapper Handler doesn't exist for specified resoure Id.
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/lib.rs:100]: Meant `resoure` -> `resource`. No rush

File: pallets/token-wrapper-handler/src/mock.rs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        //How do I know which of these I need?
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
+		Balances: pallet_balances::{Pallet, Call, Storage, Event<T>},
+		AssetRegistry: asset_registry::{Pallet, Call, Storage, Event<T>},
+		Currencies: orml_currencies::{Pallet, Call, Event<T>},
+		Tokens: orml_tokens::{Pallet, Storage, Call, Event<T>},
+		Treasury: pallet_treasury::{Pallet, Call, Storage, Config, Event<T>},
+		TokenWrapper: pallet_token_wrapper::{Pallet, Call, Storage, Event<T>},
+        TokenWrapperHandler: pallet_token_wrapper_handler::{Pallet, Call, Storage, Event<T>},
+        Bridge: pallet_bridge::<Instance1>::{Pallet, Call, Storage, Event<T>}
```
Comment (shekohex) [inline pallets/token-wrapper-handler/src/mock.rs:35]:
```suggestion
		TokenWrapperHandler: pallet_token_wrapper_handler::{Pallet, Call, Storage, Event<T>},
		Bridge: pallet_bridge::<Instance1>::{Pallet, Call, Storage, Event<T>}
```

File: pallets/token-wrapper-handler/src/mock.rs [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+type BlockNumber = u64;
+
+// Configure a mock runtime to test the pallet.
+frame_support::construct_runtime!(
+	pub enum Test where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+        //How do I know which of these I need?
```
Comment (shekohex) [inline pallets/token-wrapper-handler/src/mock.rs:26]: Which ones? the pallets?

File: pallets/token-wrapper-handler/src/mock.rs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+	type ChainIdentifier = ChainIdentifier;
+	type Event = Event;
+	type Proposal = Call;
+	type ProposalLifetime = ProposalLifetime;
+}
+
+impl Config for Test {
+	type Event = Event;
+	type BridgeOrigin = pallet_bridge::EnsureBridge<Test, BridgeInstance>;
+    type TokenWrapper = TokenWrapper; //not sure what to fill in;
```
Comment (shekohex) [inline pallets/token-wrapper-handler/src/mock.rs:194]:
```suggestion
	type TokenWrapper = TokenWrapper;
```

**Review by shekohex — CHANGES_REQUESTED** [review]
Looks good to me, a few changes regards the code formatting, nothing else.

File: pallets/asset-registry/src/lib.rs [hunk 10]
```diff
@@ -556,6 +556,14 @@ impl<T: Config> Pallet<T> {
 			},
 		)
 	}
+
+	#[allow(clippy::ptr_arg)]
```
Comment (drewstone) [inline pallets/asset-registry/src/lib.rs:560]: Do we need this here? Doesn't seem like any pointer arg is passed to this function.

File: pallets/token-wrapper-handler/src/tests.rs [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+	assert_ok!(Bridge::acknowledge_proposal(
+		Origin::signed(RELAYER_B),
+		prop_id,
+		src_chain_id,
+		*resource_id,
+		Box::new(update_proposal)
+	));
+}
+
+fn relay_remove_token_proposal(src_chain_id: u32, resource_id: &[u8; 32], prop_id: u64, name: Vec<u8>, asset_id: u32) {
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/tests.rs:92]: We can consolidate the `relay_add` and `relay_remove` by requiring them to pass in the update proposal. Otherwise it's just code duplication. Please de-duplicate and push update proposal creation to the respective test.

File: pallets/token-wrapper-handler/src/tests.rs [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+		.unwrap();
+
+		let pool_share_id = AssetRegistry::register_asset(
+			b"meme".to_vec().try_into().unwrap(),
+			AssetType::PoolShare(vec![]),
+			existential_balance.into(),
+		)
+		.unwrap();
+
+		//Setup necessary relayers/bridge functionality
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/tests.rs:147]:
```suggestion
		// Setup necessary relayers/bridge functionality
```

**Review by drewstone — CHANGES_REQUESTED** [review]
We should also always add more edge cases to the tests @saiakilesh.
- Removing token from pool without that token test
- Adding many tokens to a pool and verifying all of them
- Adding the same token to a pool and ensuring that fails or providing reason for output
- Adding a non-existent token to a pool and ensuring that fails or providing reason for output

File: pallets/asset-registry/src/lib.rs (hunk 10 above)
Comment (saiakilesh) [reply pallets/asset-registry/src/lib.rs]: fixed

File: pallets/token-wrapper-handler/src/tests.rs (hunk 11 above)
Comment (saiakilesh) [reply pallets/token-wrapper-handler/src/tests.rs]: fixed

File: pallets/token-wrapper-handler/src/tests.rs [hunk 13]
```diff
@@ -144,30 +132,88 @@ fn should_add_token() {
...
+		));
+
+		assert_eq!(AssetRegistry::contains_asset(pool_share_id, first_token_id), false);
 	})
 }
 
+/// Removing token from pool without that token test
 #[test]
-fn should_remove_token() {
+fn remove_token_not_in_pool() {
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/tests.rs:207]: `should_fail_to_remove_token_not_in_pool`

File: pallets/token-wrapper-handler/src/tests.rs [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+use pallet_bridge::types::{ProposalStatus, ProposalVotes};
+
+const TEST_THRESHOLD: u32 = 2;
+
+fn get_add_token_resource() -> Vec<u8> {
+	b"TokenWrapperHandler.execute_add_token_to_pool_share".to_vec()
+}
+
+fn get_remove_token_resource() -> Vec<u8> {
+	b"TokenWrapperHandler.execute_add_token_to_pool_share".to_vec()
```
Comment (drewstone) [inline pallets/token-wrapper-handler/src/tests.rs:16]:
```suggestion
	b"TokenWrapperHandler.execute_remove_token_to_pool_share".to_vec()
```
